src/filesystem/page_cache_bucket_allocator.h

```diff
@@ -0,0 +1,172 @@
+    return is_empty() || free_count_ == kAllocatorSize;
+  }
+  inline void evict() {
+    offset_     = 0;
+    free_count_ = 0;
+  }
+  const size_t page_size;
+
+ private:
+  uint16_t free_count_{0};
+  uint16_t offset_{0};
+  char     data_[kAllocatorSize];
+
+#ifndef NDEBUG
+  uint64_t tid_ = std::this_thread::get_id();
+#endif
+};
+
+class page_cache_bucket_allocator {
+ public:
+  struct page {
+    page(char *data, page_cache_bump_allocator *real)
+      : data_(DTHROW_IFNULL(data)), bump_(DTHROW_IFNULL(real)) {}
+
+    const char *get() const { return data_; }
+    static void dispose(page *p) { bump_->dealloc(p->get()); }
+
+   private:
+    const char *               data_;
+    page_cache_bump_allocator *bump_;
+  };
+  using ptr_t = std::lw_shared_ptr<page>;
+
+  page_cache_bucket_allocator(uint64_t min_bytes, uint64_t max_bytes)
+    : min_bytes_(min_bytes), max_bytes_(max_bytes) {
+    DLOG_THROW_IF(min_bytes_ % 4096 != 0,
+                  "incorrect min_bytes for allocator: {}", min_bytes_);
+    DLOG_THROW_IF(max_bytes_ % 4096 != 0,
+                  "incorrect min_bytes for allocator: {}", max_bytes_);
+    /*
+      XXX: Need to add cpu idler handler for this!
+
+      engine().set_idle_cpu_handler([] (reactor::work_waiting_on_reactor
+      check_for_work) {
+      return logalloc::shard_tracker().compact_on_idle(check_for_work);
+      });
+
+     */
+  }
+
+
+  uint64_t min_bytes() const { return min_bytes_; }
+  uint64_t max_bytes() const { return max_bytes_; }
+  void set_max_bytes(uint64_t new_max_bytes) { max_bytes_ = new_max_bytes; }
+
+  ptr_t alloc(size_t sz) {
+    DLOG_THROW_IF(4096 % sz != 0, "cannot allocate page size: {}", sz);
+
+
+    if (used_bytes() >= max_bytes_) { reclaim(); }
+
+    auto get_alloc = [this](auto &vec, uint64_t &idx, size_t pg_size) {
+      if (vec.is_empty()) {
+        vec.emplace_back(std::make_unique<page_cache_bump_allocator>(pg_size));
+      }
+      auto a = vec[idx].get();
+      if (a->is_full()) {
+        LOG_THROW_IF(used_bytes() >= max_bytes_,
+                     "memory full (even after reclamation) used bytes: {}",
+                     used_bytes());
+        ++idx;
+        vec.emplace_back(std::make_unique<page_cache_bump_allocator>(pg_size));
+        a = vec[idx].get();
+      };
+      return a;
+    };
+
+
+    switch (sz) {
+    case 512:
+      auto  a = get_alloc(alloc512_, alloc512_idx_, 512);
+      retun seastar::make_lw_shared<page>(a.alloc(), &a);
+    case 4096:
+      auto  a = get_alloc(alloc4096_, alloc4096_idx_, 4096);
+      retun seastar::make_lw_shared<page>(a.alloc(), &a);
+    default:
+      LOG_THROW("Could not find allocator for alignment of {}", sz);
+    }
+  }
+  inline uint64_t used_bytes() {
+    return (alloc512_.size() * kAllocatorSize)
+           + (alloc4096_.size() * kAllocatorSize);
+  }
+
+
+  void reclaim() {
+    if (used_bytes() <= max_bytes_) { return; }
+
+    auto reclaim_one = [](auto &vec) {
```

Comment:
need to decrease the index, but maybe we don't need it, we can always use .rbegin() ?